### PR TITLE
OSDOCS-8766: Add configurable connectivity check

### PR DESCRIPTION
**Pull request by jab-rh**
- https://issues.redhat.com/browse/OSDOCS-8766

Version(s): 4.16+

Issue:

Link to docs preview:

[Connectivity check source and target placement](https://75944--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/verifying-connectivity-endpoint.html#verifying-connectivity-endpoint-pod-placement)

QE review:
- [ ] QE has approved this change.

Additional information:

### CODE REVIEW

**Comment by jab-rh** [comment]
@kyrtapz, this is a first pass at documentation on this. Can you take a look when you have a chance? Thanks!

File: tmp.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+```
+    - lastTransitionTime: "2024-04-22T09:07:59Z"
+      message: No source pods available
+      reason: NoSources
+      status: "False"
+      type: NetworkDiagnosticsAvailable
+```
+
+# mode: ""   means mode: All  ?
```
Comment (kyrtapz) [inline tmp.yml:23]: https://github.com/openshift/api/pull/1800/files#diff-27a079803ccb4eb5829f5893ed44d6682a3dd99cfae4ffb3d74d35a8885c395dR223-R224

"" currently means `All` but in principal it means "sane default, the user doesn't care"

File: networking/verifying-connectivity-endpoint.adoc [hunk 2]
```diff
@@ -11,5 +11,49 @@ By reviewing the results of the health checks, you can diagnose connection probl
...
+:FeatureName: Custom configuration for connectivity check pod placement
+include::snippets/technology-preview.adoc[]
+
+You can configure the nodes on which network connectivity sources and targets run with a node selector. Additionally, you can specify permissible _tolerations_ for source and target pods. The configuration is defined in the singleton `cluster` custom resource of the `Network` API in the `config.openshift.io/v1` API group.
+
+Pod scheduling occurs after you have updated the configuration. Therefore, you must apply node labels that you intend to use in your selectors before updating the configuration. Labels applied after updating your network connectivity check pod placement are ignored.
+
+Refer to the default configuration in the following YAML:
+
+.Default configuration for connectivity source and target pods
```
Comment (kyrtapz) [inline networking/verifying-connectivity-endpoint.adoc:27]: This section is a little confusing, currently this is the default value on a fresh 4.16 cluster:
```
    networkDiagnostics:
      mode: ""
      sourcePlacement: {}
      targetPlacement: {}
```

`mode: ""` - The user doesn't care about the mode (currently defaults to `All`)
`sourcePlacement: {}` - Currently defaults to any linux worker node - nodeSelector is set to `"kubernetes.io/os": "linux"` and there are no additional tolerations.
`targetPlacement: {}` -  Currently defaults to all linux nodes - nodeSelector is set to `"kubernetes.io/os": "linux"` and it tolerates all taints.

In general empty values mean that the CNO will decide the default values and there is a chance they will change in the future.

File: tmp.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Also, TechPreview featureset is required here
+# https://docs.openshift.com/container-platform/4.15/nodes/clusters/nodes-cluster-enabling-features.html#nodes-cluster-enabling-features-cli_nodes-cluster-enabling
+
+# The nodeSelector is evaluated during pod scheduling only, that means that if you change the node labels after the pod was created it won't move.
+# For your test case this means that you have to label the node before changing the network.config settings
+
+# If no match node, the the pod will be in pending status, right ?
+# Pending means it is waiting for an available node
```
Comment (kyrtapz) [inline tmp.yml:8]: That's correct. If there are no nodes available to host he network-check-source deployment the pod will be stuck in pending until such node becomes available.

File: tmp.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        kubernetes.io/os: linux
+        node-role.kubernetes.io/infra: ""
+    targetPlacement:
+      nodeSelector:
+        kubernetes.io/os: linux
+      tolerations:
+        - operator: Exists
+status:
+
+  networkDiagnostics:
```
Comment (kyrtapz) [inline tmp.yml:56]: Please note that the `networkDiagnostics` is not a part of the status field.

File: modules/nw-pod-network-connectivity-configuration.adoc [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+$ oc edit network.config.openshift.io cluster
+----
+
+. In the text editor, update the `networkDiagnostics` stanza to specify the node selectors that you want for the source and target pods.
+
+. To commit your changes, save your changes and exit the text editor.
+
+.Verification
+
+To verify that the source and target pods are running on the intended nodes, enter the following command:
```
Comment (kyrtapz) [inline modules/nw-pod-network-connectivity-configuration.adoc:29]: I think it is also worth mentioning the `NetworkDiagnosticsAvailable` condition in `cluster/network.config.openshift.io/v1`, for example:
```
  status:
...
    conditions:
    - lastTransitionTime: "2024-05-27T08:28:39Z"
      message: ""
      reason: AsExpected
      status: "True"
      type: NetworkDiagnosticsAvailable
```

File: networking/verifying-connectivity-endpoint.adoc [hunk 6]
```diff
@@ -11,5 +11,49 @@ By reviewing the results of the health checks, you can diagnose connection probl
...
+metadata:
+  name: cluster
+spec:
+  # ...
+  networkDiagnostics:
+    mode: All
+    sourcePlacement: <1>
+      nodeSelector:
+        kubernetes.io/os: linux
+        node-role.kubernetes.io/infra: ""
```
Comment (kyrtapz) [inline networking/verifying-connectivity-endpoint.adoc:41]: `node-role.kubernetes.io/infra: ""` is not used by default.

File: modules/nw-pod-network-connectivity-configuration.adoc [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+[source,terminal]
+----
+$ oc get pods  -n openshift-network-diagnostics  -o wide
+----
+
+.Example output
+[source,text]
+----
+NAME                                    READY   STATUS    RESTARTS   AGE   IP            NODE                                        NOMINATED NODE   READINESS GATES
+network-check-source-75b88967df-mlfsn   0/1     Running   0          43s   <none>        <none>
```
Comment (kyrtapz) [inline modules/nw-pod-network-connectivity-configuration.adoc:40]: This output looks a little weird, there are no network-check-target pods.
Additionally the source pod is not ready and there is no node assigned.
Example output from a 4.16 cluster with default settings:
```
oc get po -n openshift-network-diagnostics -o wide

NAME                                    READY   STATUS    RESTARTS   AGE   IP            NODE                                                             NOMINATED NODE   READINESS GATES
network-check-source-7cdbc9858d-77bsv   1/1     Running   0          19m   10.128.2.12   pdiak-05-27-2024-xt8qh-worker-b-5f99q                            <none>           <none>
network-check-target-225wg              1/1     Running   0          19m   10.130.0.3    pdiak-05-27-2024-xt8qh-master-2.c.openshift-gce-devel.internal   <none>           <none>
network-check-target-45vgz              1/1     Running   0          19m   10.129.0.3    pdiak-05-27-2024-xt8qh-master-1.c.openshift-gce-devel.internal   <none>           <none>
network-check-target-5vlpx              1/1     Running   0          19m   10.128.0.4    pdiak-05-27-2024-xt8qh-master-0.c.openshift-gce-devel.internal   <none>           <none>
network-check-target-7qj6h              1/1     Running   0          10m   10.128.2.4    pdiak-05-27-2024-xt8qh-worker-b-5f99q                            <none>           <none>
network-check-target-f9r44              1/1     Running   0          10m   10.129.2.4    pdiak-05-27-2024-xt8qh-worker-a-w6qdk                            <none>           <none>
network-check-target-r92nl              1/1     Running   0          10m   10.131.0.3    pdiak-05-27-2024-xt8qh-worker-c-hxkd4                            <none>           <none>
```

File: tmp.yml [hunk 8]
```diff
@@ -0,0 +1,66 @@
+# Also, TechPreview featureset is required here
```
Comment (kyrtapz) [inline tmp.yml:1]: The feature is already in GA in 4.16

**Comment by jab-rh** [comment]
@kyrtapz, oops, `tmp.yaml` was not supposed to be included in the PR. I'll remove it.

Thanks!

**Comment by jab-rh** [comment]
@kyrtapz, I've updated the draft based on your feedback, thanks!

File: modules/nw-pod-network-connectivity-configuration.adoc [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+[source,terminal]
+----
+$ oc get pods -n openshift-network-diagnostics -o wide
+----
+
+.Example output
+[source,text]
+----
+NAME                                    READY   STATUS    RESTARTS   AGE   IP            NODE                                                             NOMINATED NODE   READINESS GATES
+network-check-source-7cdbc9858d-77bsv   1/1     Running   0          19m   10.128.2.12   pdiak-05-27-2024-xt8qh-worker-b-5f99q                            <none>           <none>
```
Comment (kyrtapz) [inline modules/nw-pod-network-connectivity-configuration.adoc:40]: NIT: Can we alter the node names so the are more generic and do not use `master`?
For example:
```
oc get po -n openshift-network-diagnostics -o wide                                                                                                                                                                                                                                                                                           
NAME                                    READY   STATUS    RESTARTS   AGE     IP           NODE                                        NOMINATED NODE   READINESS GATES
network-check-source-84c69dbd6b-p8f7n   1/1     Running   0          9h      10.131.0.8   ip-10-0-40-197.us-east-2.compute.internal   <none>           <none>
network-check-target-46pct              1/1     Running   0          9h      10.131.0.6   ip-10-0-40-197.us-east-2.compute.internal   <none>           <none>
network-check-target-8kwgf              1/1     Running   0          9h      10.128.2.4   ip-10-0-95-74.us-east-2.compute.internal    <none>           <none>
network-check-target-jc6n7              1/1     Running   0          9h      10.129.2.4   ip-10-0-21-151.us-east-2.compute.internal   <none>           <none>
network-check-target-lvwnn              1/1     Running   0          9h      10.128.0.7   ip-10-0-17-129.us-east-2.compute.internal   <none>           <none>
network-check-target-nslvj              1/1     Running   0          9h      10.130.0.7   ip-10-0-89-148.us-east-2.compute.internal   <none>           <none>
network-check-target-z2sfx              1/1     Running   0          9h      10.129.0.4   ip-10-0-60-253.us-east-2.compute.internal   <none>           <none>

```

**Review by aldiazRH — COMMENTED** [review]
LGTM! Small nits here and there that are pretty optional.

**Comment by jab-rh** [comment]
@aldiazRH, thanks!

**Comment by jab-rh** [comment]
@mffiedler, can you take a look at this docs draft when you some time for the 4.16 release? Thanks!

**Comment by kyrtapz** [comment]
LGTM
Thanks @jab-rh!

**Comment by anuragthehatter** [comment]
cc @huiran0826

**Pull request merged**